[WIP] Patch script updates for vim's GitHub migration

jwhitley opened the pull request:
Upstream Vim is migrating to GitHub from Google Code, due to the latter's impending shutdown.  This PR updates the patch script for that move.

jwhitley committed:
Patch script updates for vim's GitHub migration

Upstream Vim is migrating to GitHub from Google Code, due to the
latter's impending shutdown.  This updates the patch script for that
move.

jwhitley commented:
At this writing, this is incomplete.  I'll finish it up shortly, but in any event it should probably not land until Vim's migration to the new repo is considered final.

justinmk commented:
:clap:

tarruda commented:
:+1:

Pyrohh commented:
> At this writing, this is incomplete. I'll finish it up shortly, but in any event it should probably not land until Vim's migration to the new repo is considered final.

Okey dokey, I'll tag this so we don't forget.

jwhitley committed:
Update functions for vim's git migration

git_vim_sources(), assign_commit_details(), and get_vim_patch() are
updated, but not yet tested.

jwhitley commented:
Remaining status checklist:

- [ ] Update `list_vim_patches()` (grok & port those gnarly hg `--template` expressions)
- [ ] Add code to handle $VIM_SOURCE_DIR existing, but being an old hg repo.

justinmk commented:
> handle $VIM_SOURCE_DIR existing, but being an old hg repo

Don't bother, unless you just want to show an error that advises the user to nuke it.

justinmk commented on scripts/vim-patch.sh:
```diff
@@ -36,17 +36,17 @@ check_executable() {
     echo "Updating Vim sources in '${VIM_SOURCE_DIR}'."
     cd "${VIM_SOURCE_DIR}"
-    hg pull --update &&
+    git pull &&
```
Should probably be a `git fetch --all`. It doesn't really matter what is actually checked out in the workspace, the script can pull from `origin/master`.

justinmk commented on scripts/vim-patch.sh:
```diff
@@ -64,17 +64,17 @@ assign_commit_details() {
+    vim_commit="v${1}"
     local strip_commit_line=true
-    vim_commit_url="https://github.com/vim/vim/commit/${vim_commit}"
+    vim_commit_url="https://github.com/vim/vim/tree/${vim_commit}"
```
`commit` works here too.

jwhitley replied:
Not in my testing.  In this code, `$vim_commit` will be a tag ref like `v7.4.723`.  E.g. compare:

https://github.com/vim/vim/commit/v7.4.723

to 

https://github.com/vim/vim/tree/v7.4.723

justinmk replied:
Yeah that's weird, it used to work on the old tryout repo. Maybe github changed something.

fwalch replied:
Not sure why this doesn't work with the Vim repo, e.g. https://github.com/neovim/neovim/commit/nightly still works fine. But maybe it would be better to just use the commit hash for the URL, same as for non-tagged commits/runtime updates? I think it's very convenient to immediately see the changes with one click.

justinmk replied:
Agreed, sounds good.

jwhitley commented:
> Don't bother, unless you just want to show an error that advises the user to nuke it.

I will definitely do this; it's a trivial change and otherwise things will break weirdly/badly when `.vim-src` already exists but is an hg repo.

jwhitley commented:
Wow, it looks like vim rewrote their hg tag history... the `v7-x-yyy` tag forms the script uses no longer work, it's now `v7.x.yyy`.

justinmk commented:
>  it looks like vim rewrote their hg tag history

That's correct. I've asked Bram to keep the old tags, at least in the google code repo. Not sure if that will happen.

jwhitley commented:
FWIW, the old `.hgtags` entires `v7-x-yyy` still exist in the file but have deletion records now.  That latter was unnecessary, and is water under the bridge unless someone reverted that change.  What I don't understand is why *any* purely conversion-related work was undertaken in the public repository vs. a local clone.

It looks like the GitHub vim repo is now official (?) per [this message from Bram](https://groups.google.com/d/msg/vim_dev/Io5A_Zir--k/zTqV3ZKUBQAJ).

The major piece of work left is just sorting out how to handle the work done by those `hg log --template [...]` commands.  There's no even proximate equivalent in git for the programmatic template options, so no simple translation is possible.  I started yesterday taking a look at the actual effect of those statements to work out a reasonable git-centric way of producing the same result, but need to press on and get that finished.

jwhitley committed:
Last hg conversions done

justinmk commented:
Template feature of hg does look nice. 

The usage in `assign_commit_details` is probably equivalent to

    git log -1 --format='%B'

Regarding the usages in `list_vim_patches`: `"Added tag"` does not need to be checked anymore, it was removed during the conversion (in hg they were separate noise commits). `vim_commits` appears to be a list of tags since 442, so it now becomes:

    local vim_commits=$(git tag --contains v7.4.442)

I don't understand why `vim_old_commits` is a separate variable. Seems like `v7.4.442` could just be switched to `v7.4.161`.

jwhitley committed:
Verify that VIM_SOURCE_DIR is a git repo

This may be a leftover hg repo from before vim's git migration.

justinmk commented on scripts/vim-patch.sh:
```diff
@@ -36,17 +36,22 @@ check_executable() {
   if [[ ! -d ${VIM_SOURCE_DIR} ]]; then
     echo "Cloning Vim sources into '${VIM_SOURCE_DIR}'."
-    hg clone https://code.google.com/p/vim "${VIM_SOURCE_DIR}"
+    git clone https://github.com/vim/vim.git "${VIM_SOURCE_DIR}"
```
`git clone --depth=1000` is probably reasonable. Not sure if it saves much space though.

jwhitley replied:
It saves ~20M space, and it almost halves the clone time.  Added in next commit.

jwhitley commented:
This is now code-complete, but needs testing.  It's ready for full review.  I've given `vim-patch.sh -l` a light tire-kicking.  Output seems plausible, but I haven't verified it.  It's late here, so I'll run it through its paces vs. the hg version tomorrow AM.

justinmk commented on scripts/vim-patch.sh:
```diff
@@ -162,22 +167,17 @@ list_vim_patches() {
-    hg log --removed --template='{if(startswith("Added tag",
-      firstline(desc)),
-      "{latesttag}\n")}' -r v7-4-442:v7-4-161)
+    git log --pretty='tformat:%D' v7.4.160..v7.4.442 | awk '{ if ($1 != "") print($2); }')
```
`git log --pretty='format:%D'` just prints `%D` for me (and `tformat` doesn't work either)`. (git 1.9.1)

jwhitley replied:
:imp:  I'm on git 2.5.0, so I've got all the goodies.  I'll check an older man page.

jwhitley replied:
1.9.1 is supposed to support tformat according to its man page, so I'm not sure what's up with that.

jwhitley replied:
Sigh, I just need to port the awk stuff to handle the parens from `%d`, which exists in 1.9.1.

jwhitley committed:
Limit depth of upstream vim clone

This about halves the amount of time it takes to clone, and saves ~25%
of the full repo space.

jwhitley committed:
Update git log formats to use %d vs %D

git 1.9.1, as on Ubuntu 14.04 LTS and elsewhere, doesn't yet implement
the %D log format.

justinmk commented on scripts/vim-patch.sh:
```diff
@@ -162,22 +167,17 @@ list_vim_patches() {
-      "{if(startswith(\"updated for version\", firstline(desc)),
-        \"\",
-        \"{node}\n\")}")}' -r tip:v7-4-442)
+    git log --pretty='tformat:%H|%d' v7.4.442..HEAD | awk -f ${NEOVIM_SOURCE_DIR}/scripts/vim-patch-helper.awk)
```
Not sure if you saw https://github.com/neovim/neovim/pull/3193#issuecomment-133289081 .

What does this list look like? Is it different than `git tag --contains v7.4.442` ?

justinmk commented on scripts/vim-patch.sh:
```diff
@@ -162,22 +167,17 @@ list_vim_patches() {
-    hg log --removed --template='{if(startswith("Added tag",
-      firstline(desc)),
-      "{latesttag}\n")}' -r v7-4-442:v7-4-161)
+    git log --pretty='tformat:%d' v7.4.160..v7.4.442 | awk '{ if ($1 != "") { gsub(/[()]/, "", $2); print($2); } }')
```
What does this list look like?

jwhitley replied:
See https://gist.github.com/e26aba8e38f56d02a6f8 the output of this and the other similar git log line.

For this line in particular, IIRC the point was that with the runtime reintegration, the tests for whether non-patch runtime revs had been integrated didn't really work.  So for the older revs the idea was to extract only the tag/patch names and leave off pulling the non-patch commit hashes.

justinmk replied:
> extract only the tag/patch names 

Sorry if I'm being dense, but if I am not mistaken, that is what `git tag --contains v7.4.442` does.

jwhitley commented:
So some bad news, which is now quite apparent.  Have a look at this gist, which contains the output of `scripts/vim-patch.sh -l` against the old hg vim repo (as `vim-patch-l.hg.txt`) and the new git vim repo (as `vim-patch-l-git.txt`)

https://gist.github.com/anonymous/54e3b75af9d92cf1ecb3

The test for whether a named patch rev (v7.x.yyy) has been included are working, but the test for whether an unnamed (id'ed by commit hash) rev is included in neovim is busted: it's trying to compare the *hg* hashes in neovim's history against the *git* commit ids in the new repo.  This means that integrated runtime and other patches from the hg history aren't being detected as merged anymore in the git history.

AFAICT, there's no way to reconcile these.  In order for this script to continue working, non-tagged revs prior to the git changeover will have to continue to be compared against the hg repo.  I.e. it'll have to pull *two* copies of vim upstream, the hg version (for old stuff) and the git version (for new stuff). :sob:

I hope/assume that the Google Code repo will actually remain extant, in read-only mode.  Otherwise a legacy mirror of it will have to be hosted somewhere, at least until all hg-era changes are ported over.

jwhitley commented:
My current thinking is that I'll look into refactoring this to preserve the hg logic (used only for legacy commits), and add the git logic in a way such that the hg stuff can easily be retired if/when all the hg-era changes have been ported to neovim.

justinmk commented:
> non-tagged revs prior to the git changeover will have to continue to be compared against the hg repo.

No, we can just hard-code the git SHAs that correspond to the old hg revs.

fwalch commented:
I think runtime patches have been merged in order, so you should be able to just restrict the range of commits that are checked (to after the last merged runtime patch) and just compare git hashes.

justinmk commented on scripts/vim-patch.sh:
```diff
@@ -253,9 +253,9 @@ review_pr() {
   echo "${neovim_patch}" > "${NEOVIM_SOURCE_DIR}/n${base_name}.patch"
   echo "✔ Saved full pull request commit details to '${NEOVIM_SOURCE_DIR}/n${base_name}.patch'."
-  hg diff --show-function --git --change "${vim_commit}" > "${NEOVIM_SOURCE_DIR}/${base_name}.diff"
+  git diff "${vim_commit}^\!" > "${NEOVIM_SOURCE_DIR}/${base_name}.diff"
```
What does `^\!` do?

jwhitley replied:
Per `[gitrevisions(7)](http://git-scm.com/docs/gitrevisions)`, that means "the range with only this commit in it" (the meaning is slightly different than that; see the man page).  It's less cumbersome than, e.g. something like "HEAD..HEAD".

justinmk commented:
Merged with some changes. If I messed something up let me know.

jwhitley commented:
Thanks all.  Sorry to have dropped off there; life intervened the past few days.  I'll review @justinmk's changes for good measure.  I also agree with @fwalch's suggestions; great idea and keeps things simple.  Presumably this means a separate list of git hashes for the as-yet-unmerged runtime patches will need to be added in the repo.
